### ISSUE TITLE
Recovered from panic in server

### ISSUE THREAD

**Issue by pytimer** [coredns/coredns#3735]
**What happened**:

I have two kubernetes cluster and use coredns with hostNetwork, i found one of cluster coredns pod output `[ERROR] Recovered from panic in server: "dns://:1553"` every few seconds. But the dns resolve is ok.

**What you expected to happen**:

coredns no error logs.

**How to reproduce it (as minimally and precisely as possible)**:

I don't know why have this problem, so i can't reproduce it.

**Anything else we need to know?**:

**Environment**:

- the version of CoreDNS: 1.6.5
- Corefile:
```yaml
apiVersion: v1
data:
  Corefile: |
    .:1553 {
        errors
        health :9180 {
           lameduck 5s
        }
        ready :9181
        kubernetes cluster.local in-addr.arpa ip6.arpa {
           pods insecure
           fallthrough in-addr.arpa ip6.arpa
           ttl 30
        }
        prometheus :9153
        forward . /etc/resolv.conf
        cache 30
        loop
        reload
        loadbalance
    }
kind: ConfigMap
metadata:
  name: coredns
  namespace: kube-system
```
- logs, if applicable:
```sh
[INFO] plugin/ready: Still waiting on: "kubernetes"
.:1553
[INFO] plugin/reload: Running configuration MD5 = ea45c18bb97e6b6409eba23a1166b357
CoreDNS-1.6.5
linux/amd64, go1.13.4, c2fd1b2
[ERROR] Recovered from panic in server: "dns://:1553"
[INFO] plugin/ready: Still waiting on: "kubernetes"
[ERROR] Recovered from panic in server: "dns://:1553"
[ERROR] Recovered from panic in server: "dns://:1553"
[ERROR] Recovered from panic in server: "dns://:1553"
[ERROR] Recovered from panic in server: "dns://:1553"
[ERROR] Recovered from panic in server: "dns://:1553"
[ERROR] Recovered from panic in server: "dns://:1553"
[ERROR] Recovered from panic in server: "dns://:1553"
[ERROR] Recovered from panic in server: "dns://:1553"
```
- OS (e.g: `cat /etc/os-release`):
```sh
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```
- Others:

coredns.yaml

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    k8s-app: kube-dns
  name: coredns
  namespace: kube-system
spec:
  replicas: 1
  selector:
    matchLabels:
      k8s-app: kube-dns
  strategy:
    type: Recreate
  template:
    metadata:
      creationTimestamp: null
      labels:
        k8s-app: kube-dns
    spec:
      affinity:
        nodeAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
            nodeSelectorTerms:
            - matchExpressions:
              - key: node-role.kubernetes.io/master
                operator: Exists
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: k8s-app
                operator: In
                values:
                - kube-dns
            topologyKey: kubernetes.io/hostname
      containers:
      - args:
        - -conf
        - /etc/coredns/Corefile
        image: docker.hikcloud:30001/k8ss/coredns:1.6.5
        imagePullPolicy: IfNotPresent
        livenessProbe:
          failureThreshold: 5
          httpGet:
            path: /health
            port: 9180
            scheme: HTTP
          initialDelaySeconds: 60
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 5
        name: coredns
        ports:
        - containerPort: 1553
          hostPort: 1553
          name: dns
          protocol: UDP
        - containerPort: 1553
          hostPort: 1553
          name: dns-tcp
          protocol: TCP
        - containerPort: 9153
          hostPort: 9153
          name: metrics
          protocol: TCP
        readinessProbe:
          failureThreshold: 3
          httpGet:
            path: /ready
            port: 9181
            scheme: HTTP
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 1
        resources:
          limits:
            memory: 170Mi
          requests:
            cpu: 100m
            memory: 70Mi
        securityContext:
          allowPrivilegeEscalation: false
          capabilities:
            add:
            - NET_BIND_SERVICE
            drop:
            - all
          readOnlyRootFilesystem: true
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /etc/coredns
          name: config-volume
          readOnly: true
      dnsPolicy: Default
      hostNetwork: true
      nodeSelector:
        beta.kubernetes.io/os: linux
      priorityClassName: system-cluster-critical
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      serviceAccount: coredns
      serviceAccountName: coredns
      terminationGracePeriodSeconds: 30
      tolerations:
      - key: CriticalAddonsOnly
        operator: Exists
      - effect: NoSchedule
        key: node-role.kubernetes.io/master
      - effect: NoSchedule
        key: node.kubernetes.io/not-ready
      volumes:
      - configMap:
          defaultMode: 420
          items:
          - key: Corefile
            path: Corefile
          name: coredns
        name: config-volume
```

**Comment by rdrozhdzh** [comment]
You can add `debug` directive to your corefile. This way your coredns will print stack trace to the log before crashing/restarting which may give some clue to the reason of failure

**Comment by chrisohaver** [comment]
> You can add `debug` directive to your corefile. This way your coredns will print stack trace to the log before crashing/restarting which may give some clue to the reason of failure

When using debug to catch panics, you'll need to also to remove `errors` from the Corefile, (the `errors` plugin negates `debug` in this regard).

**Comment by chrisohaver** [comment]
Also curious of the timing of the panics.  
Can you add time stamps to the output? `kubectl logs --timestamps ...`

**Comment by pytimer** [comment]
> > You can add `debug` directive to your corefile. This way your coredns will print stack trace to the log before crashing/restarting which may give some clue to the reason of failure
> 
> When using debug to catch panics, you'll need to also to remove `errors` from the Corefile, (the `errors` plugin negates `debug` in this regard).

I using `debug`, i found the panic because of `forward` plugin.
Coredns load `/etc/resolv.conf` but no nameserver in this file, so report panic.  I add nameserver in `/etc/resolv.conf`, this error is gone. I don't know why?

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Recovered fro..." ]
>> corefile
>
>I using `debug`, i found the panic because of `forward` plugin.
>Coredns load `/etc/resolv.conf` but no nameserver in this file, so report panic.  I add nameserver in `/etc/resolv.conf`, this error is gone. I don't know why?

that's bug

/plugin forward
/label bug

**Comment by pytimer** [comment]
The code panic : https://github.com/coredns/coredns/blob/v1.6.5/plugin/forward/forward.go#L85

**Comment by chrisohaver** [comment]
If we make `parse.HostPortOrFile()` return an error when there are no resulting servers this would get caught at startup.  Looking at the existing (in tree) places that use it now, looks safe to do.  Or if we don't want that to be considered an error, just check for 0 length result in `forward.parseStanza()`.

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "Re: [coredns/coredns] Recovered fro..." ]
>If we make `parse.HostPortOrFile()` return an error when there are no resulting servers this would get caught at startup.  Looking at the existing (in tree) places that use it now, looks safe to do.  Or if we don't want that to be considered an error, just check for 0 length result in `forward.parseStanza()`.

explcitly checking for 0 is better I think, although now all callers need to do this...

**Comment by miekg** [comment]
closed via #3742
